sw/host/provisioning/orchestrator/src/device_id.py

```diff
@@ -0,0 +1,158 @@
+        """Validates this object's attributes."""
+        # Validate year.
+        if not 0 <= self.year <= 9:
+            raise ValueError("Year ({}) must be in [0, 9].".format(self.year))
+        # Validate week.
+        if not 0 <= self.week <= 51:
+            raise ValueError("Week ({}) must be in [0, 51].".format(self.week))
+        # Validate lot.
+        if not 0 <= self.lot <= 999:
+            raise ValueError("Lot ({}) must be in [0, 999].".format(self.lot))
+        # Validate wafer.
+        if not 0 <= self.wafer <= 99:
+            raise ValueError("Wafer ({}) must be in [0, 99].".format(
+                self.wafer))
+        # Validate wafer X coordinate.
+        if not 0 <= self.wafer_x_coord <= 999:
+            raise ValueError(
+                "Wafer X coordinate ({}) must be in [0, 999].".format(
+                    self.wafer_x_coord))
+        # Validate wafer Y coordinate.
+        if not 0 <= self.wafer_y_coord <= 999:
+            raise ValueError(
+                "Wafer Y coordinate ({}) must be in [0, 999].".format(
+                    self.wafer_y_coord))
+
+    def to_int(self) -> int:
+        din = 0
+        din |= self.wafer_y_coord << 44
+        din |= self.wafer_x_coord << 32
+        din |= self.wafer << 24
+        din |= self.lot << 12
+        din |= self.week << 4
+        din |= self.year
+        return din
+
+
+class DeviceId():
+    """An OpenTitan device ID.
+
+    This object can generate different encodings of the device ID, including:
+      - hex string, and
+      - int.
+
+    Attributes:
+        si_creator_id: A 16-bit number assigned by the OpenTitan project.
+        product_id: A 16-bit number assigned by the OpenTitan project.
+        din: A 64-bit unique identifier.
+        crc32: A CRC32 over the above 96-bits.
+        package_id: A 16-bit number indicating which package the chip is in.
+        sku_id: A 32-bit string indicating the SKU the chip was provisioned for.
+    """
+
+    def __init__(self, sku_config: SkuConfig, din: DeviceIdentificationNumber):
+        # Save string keys for easier printing.
+        self._product = sku_config.product
+        self._si_creator = sku_config.si_creator
+        self._package = sku_config.package
+        self._sku = sku_config.name
+
+        # Build HW origin with:
+        # - 16 bits SiliconCreator ID
+        # - 16 bits Product ID
+        self.si_creator_id = sku_config.si_creator_id
+        self.product_id = sku_config.product_id
+        self._hw_origin = bytes_to_int(
+            struct.pack("<HH", self.si_creator_id, self.product_id))
+
+        # Build Device Identification Number with:
+        # - Year
+        # - Week
+        # - Lot #
+        # - Wafer #
+        # - Wafer X coord.
+        # - Wafer Y coord.
+        self.din = din
+
+        # Build CRC32 over HW origin + DIN.
+        self.crc32 = binascii.crc32(
+            struct.pack("<IQ", self._hw_origin, self.din.to_int()))
```

Comment:
Did we agree to do away with the CRC32 in the device id since the OTP partition already has an integrity check?

Reply:
Fixed in #25167.